src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -1668,6 +1668,11 @@ protected void AfterMembersChecks(BindingDiagnosticBag diagnostics)
             CheckForUnmatchedOperators(diagnostics);
             CheckForRequiredMemberAttribute(diagnostics);
 
+            if (IsScriptClass || IsSubmissionClass)
+            {
+                ReportRequiredMembers(diagnostics);
+            }
```

Comment:
https://github.com/dotnet/roslyn/blob/e7a1a087d58366883d2b3995b8cd071db1a1ee6f/src/Compilers/CSharp/Portable/Declarations/DeclarationKind.cs#L22-L24

What `ImplicitClass` is for? Do we need to check for `IsImplicitClass`? If yes, does it make sense to report the diagnostic in [`ImplicitNamedTypeSymbol`](https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/Symbols/Source/ImplicitNamedTypeSymbol.cs) instead?  (by overriding AfterMembersCompletedChecks or making AfterMembersChecks virtual and override it)

Reply:
`ImplicitClass` is for top-level statements. Those get separate errors for declaring members at all, it's not necessary to report errors for required members specifically here.

Reply:
@333fred I'm confused. `ImplicitClass` seems to be a very old thing per Git history:

![image](https://user-images.githubusercontent.com/31348972/173680689-4e0ccc55-76fd-4511-aaf1-7f54a311b0fc.png)

Reply:
@Youssef1313 Looks like a problem with GitHub. The git blame in VS shows:
![image](https://user-images.githubusercontent.com/12466233/173689036-9bee29d8-33ce-4aec-8967-612e61d05d41.png)

Reply:
@jcouv Youssef is correct. You're reading the git blame wrong: Aleksey added a _comma_ in 2020.

However, it doesn't change the answer. The C# compiler has long supported giving good errors when members are declared in namespaces, or without a namespace at all. These are all error scenarios, and ImplicitClass is used for them.